### ISSUE TITLE
Proving user feedback for track uploading errors

### ISSUE THREAD

**Issue by SebaDro** [enviroCar/enviroCar-app#870]
**Description**
If an error occurs during upload of **all** local tracks, the user does not get a proper feedback about the reason for the upload failure. The error message just shows "An error occurred during the upload process.". A more meaningful message text would be necessary to assess if the failure is effected by a misuse of the app, an app error or a server failure. In addition, the error is not logged, thus it does not appear within the log report.

**Branches**
master, develop

**How to reproduce**
Record a small track with and finish recording. Go to the LocalTrackFragment, turn off internet and click the button for uploading all tracks => the app shows "An error occurred during the upload process."

**How to fix**
Adjust error handling here: https://github.com/enviroCar/enviroCar-app/blob/98403d7bc73e246bdae761e515c78c3e688a0dfa/org.envirocar.app/src/org/envirocar/app/views/tracklist/TrackListLocalCardFragment.java#L530-L536

Add error logging and enhance the error handling. Maybe, distinguishing the error type such as for single track upload errors would be useful: https://github.com/enviroCar/enviroCar-app/blob/98403d7bc73e246bdae761e515c78c3e688a0dfa/org.envirocar.app/src/org/envirocar/app/views/tracklist/TrackListLocalCardFragment.java#L390-L418

**Comment by cdhiraj40** [comment]
@SebaDro I think its the same issue as #857

**Comment by SebaDro** [comment]
Hey @cdhiraj40. Yes, you're right. In parts, the PR adresses this issue. However, it only focusses on the internet connection problem. A proper handling of other error types is still pending. So, I think it's ok to leave both issues open.

**Comment by cdhiraj40** [comment]
> Hey @cdhiraj40. Yes, you're right. In parts, the PR adresses this issue. However, it only focusses on the internet connection problem. A proper handling of other error types is still pending. So, I think it's ok to leave both issues open.

yes I totally agree, I will look into the issue and see why it's not working, thanks :)

**Comment by cdhiraj40** [comment]
@SebaDro  I think its happening cause here 
```        public void onError(Throwable e) {
```
its throwing ``NullPointerException`` 

detailed version : ```java.lang.NullPointerException cannot be cast to org.envirocar.core.exception.TrackUploadException

**Comment by SebaDro** [comment]
Yes, I noticed the `NullPointerException`, too. If you follow the stacktrace, you will find, that the exception occurs inside the `AgreementManager`: https://github.com/enviroCar/enviroCar-app/blob/98403d7bc73e246bdae761e515c78c3e688a0dfa/org.envirocar.app/src/org/envirocar/app/handler/agreement/AgreementManager.java#L122

It seems, there are some implementions missing for the [CacheTermsOfUseDAO](https://github.com/enviroCar/enviroCar-app/blob/master/org.envirocar.remote/src/main/java/org/envirocar/remote/dao/CacheTermsOfUseDAO.java).

**Comment by cdhiraj40** [comment]
> Yes, I noticed the `NullPointerException`, too. If you follow the stacktrace, you will find, that the exception occurs inside the `AgreementManager` when it comes to check the terms of use:
> 
> https://github.com/enviroCar/enviroCar-app/blob/98403d7bc73e246bdae761e515c78c3e688a0dfa/org.envirocar.app/src/org/envirocar/app/handler/agreement/AgreementManager.java#L119-L139
> 
> It seems, there are some implementions missing for the [CacheTermsOfUseDAO](https://github.com/enviroCar/enviroCar-app/blob/master/org.envirocar.remote/src/main/java/org/envirocar/remote/dao/CacheTermsOfUseDAO.java).
> 
> However, I don't think there is a necessity to check the terms of use, if there is no internet. I'd prefer a "fail fast" solution e.g. checking internet connection and showing an error dialog without trying to upload a track.

I think I did try that in #857, maybe we can make functions and use it?

**Comment by SebaDro** [comment]
Have been fixed in parts with #874. However, the app still does not inform the user about the cause for a failing track upload.

**Comment by cdhiraj40** [comment]
@SebaDro sir is this issue from the server side? if not then we can try handling one by one and check why its not working properly

**Comment by SebaDro** [comment]
I'm not sure if the server responds with a meaningful error message for each track. First step would be to debug the app and analyze the exceptions.
